### PR TITLE
Added sync mode

### PR DESCRIPTION
**Pull request by mcollina**
This is mainly to solve https://github.com/pinojs/pino/issues/542, however I found out that it gives us a  20% speed bump when logging things with sync mode in pino benchmarks.

Benchmarks in pino with sonic-boom master:

```
$ node benchmarks/basic.bench.js
benchPino*10000: 162.391ms
benchPinoExtreme*10000: 108.874ms
benchPinoNodeStream*10000: 245.955ms
```

Benchmarks in pino with this PR, enabling sync mode for both destination and extreme:

```
$ node benchmarks/basic.bench.js
benchPino*10000: 147.822ms
benchPinoExtreme*10000: 83.608ms
benchPinoNodeStream*10000: 243.117ms
```

cc @jsumners @davidmarkclements

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,13 +4,14 @@ Extremely fast utf8-only stream implementation to write to files and
 file descriptors.
 
 This implementation is partial, but support backpressure and `.pipe()` in is here.
-However, it is 20x faster than Node Core `fs.createWriteStream()`:
+However, it is 2-3x faster than Node Core `fs.createWriteStream()`:
```
Comment (davidmarkclements) [inline README.md]: has the number dropped because Node's gotten faster?

File: README.md [hunk 2]
```diff
@@ -52,6 +53,9 @@ The first argument can be:
 The second argument is the minimum length of the internal buffer that is
 required before flushing.
 
+The third argument is a flag that make SonicBoom work synchronously or
```
Comment (davidmarkclements) [inline README.md]:
```suggestion
The third argument is a flag that, when true, causes `SonicBoom` to perform synchronous writes.
```

File: README.md [hunk 3]
```diff
@@ -52,6 +53,9 @@ The first argument can be:
...
+not.
```
Comment (davidmarkclements) [inline README.md]:
```suggestion

```

**Comment by davidmarkclements** [comment]
is this just a case of optimizing for benchmarks though?

Does it show improvement in pino-http / hapi-pino benchmarks?

File: README.md (hunk 1 above)
Comment (mcollina) [reply README.md]: Yes! 🚀 

I also adjusted the benchmark to be more realistic

**Comment by mcollina** [comment]
> is this just a case of optimizing for benchmarks though?

It seems so, yes. In fact, we are slowing things down with pino-http for example by 30% when using sync mode. It's still 20% faster than a standard Node.js stream.

> Does it show improvement in pino-http / hapi-pino benchmarks?

Not at all.

---

TL;DR

I will assemble a PR to fix https://github.com/pinojs/pino/issues/542 when using the standard destination. We will lose throughput, but I think that type of behavior should be opt-in.

**Comment by mcollina** [comment]
My plan for pino is to have:

1) sync mode if the destination is stdout, this solves https://github.com/pinojs/pino/issues/542.
2) retain extreme mode as is.
3) provide an option object to `pino.destination()` and `pino.extreme()`, so that users could in theory pass in `sync: true` or  `sync: false`. I plan to make the default for `pino.destination()` synchronous for safety reasons.

This should match the behavior of Node core as much as possible.

cc @jsumners @davidmarkclements

**Comment by davidmarkclements** [comment]
1. 👍 
2. 👍
3. by definition, destination should be maximum safety, extreme should be maximum performance. From an API perspective I don't see much benefit to allowing sync true|false
